## 1.2 Exercise: Exploring a Pandas Data Frame

In [1]:
# First I will import some needed libraries
import numpy as np
import pandas as pd

## Load the dataset as a Pandas data frame.

First I will import CSV File Video_Games_Sales_as_at_22_Dec_2016.csv through the use of Pandas.

In [2]:
Video_Game_Sales = pd.read_csv("Video_Games_Sales_as_at_22_Dec_2016.csv")
Video_Game_Sales

Name Platform  Year_of_Release         Genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16714  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16715               LMA Manager 2007     X360           2006.0        Sports   
16716        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16717               Spirits & Spells      GBA           2003.0      Platform   
16718            Winning Post 8 2016      PSV           2016.0    Simulation   

          Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
0          Nintendo     41.36     28.96      3.77         8.45         82.53   
1          Nintendo     29.08      3.58      6.81         0.77         40.24   
2          Nintendo     15.68     12.76      3.79         3.29         35.52   
3          Nintendo     15.61     10.93      3.28         2.95         32.77   
4          Nintendo     11.27      8.89     10.22         1.00         31.37   
...             ...       ...       ...       ...          ...           ...   
16714    Tecmo Koei      0.00      0.00      0.01         0.00          0.01   
16715   Codemasters      0.00      0.01      0.00         0.00          0.01   
16716  Idea Factory      0.00      0.00      0.01         0.00          0.01   
16717       Wanadoo      0.01      0.00      0.00         0.00          0.01   
16718    Tecmo Koei      0.00      0.00      0.01         0.00          0.01   

       Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
0              76.0          51.0          8       322.0  Nintendo      E  
1               NaN           NaN        NaN         NaN       NaN    NaN  
2              82.0          73.0        8.3       709.0  Nintendo      E  
3              80.0          73.0          8       192.0  Nintendo      E  
4               NaN           NaN        NaN         NaN       NaN    NaN  
...             ...           ...        ...         ...       ...    ...  
16714           NaN           NaN        NaN         NaN       NaN    NaN  
16715           NaN           NaN        NaN         NaN       NaN    NaN  
16716           NaN           NaN        NaN         NaN       NaN    NaN  
16717           NaN           NaN        NaN         NaN       NaN    NaN  
16718           NaN           NaN        NaN         NaN       NaN    NaN  

[16719 rows x 16 columns]

## Display the first ten rows of data.

Next I will use "head" to display the first ten rows of data by insirting ten in the head brackets as seen below:

In [3]:
Video_Game_Sales.head(10)

Name Platform  Year_of_Release         Genre  \
0                 Wii Sports      Wii           2006.0        Sports   
1          Super Mario Bros.      NES           1985.0      Platform   
2             Mario Kart Wii      Wii           2008.0        Racing   
3          Wii Sports Resort      Wii           2009.0        Sports   
4   Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
5                     Tetris       GB           1989.0        Puzzle   
6      New Super Mario Bros.       DS           2006.0      Platform   
7                   Wii Play      Wii           2006.0          Misc   
8  New Super Mario Bros. Wii      Wii           2009.0      Platform   
9                  Duck Hunt      NES           1984.0       Shooter   

  Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
0  Nintendo     41.36     28.96      3.77         8.45         82.53   
1  Nintendo     29.08      3.58      6.81         0.77         40.24   
2  Nintendo     15.68     12.76      3.79         3.29         35.52   
3  Nintendo     15.61     10.93      3.28         2.95         32.77   
4  Nintendo     11.27      8.89     10.22         1.00         31.37   
5  Nintendo     23.20      2.26      4.22         0.58         30.26   
6  Nintendo     11.28      9.14      6.50         2.88         29.80   
7  Nintendo     13.96      9.18      2.93         2.84         28.92   
8  Nintendo     14.44      6.94      4.70         2.24         28.32   
9  Nintendo     26.93      0.63      0.28         0.47         28.31   

   Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
0          76.0          51.0          8       322.0  Nintendo      E  
1           NaN           NaN        NaN         NaN       NaN    NaN  
2          82.0          73.0        8.3       709.0  Nintendo      E  
3          80.0          73.0          8       192.0  Nintendo      E  
4           NaN           NaN        NaN         NaN       NaN    NaN  
5           NaN           NaN        NaN         NaN       NaN    NaN  
6          89.0          65.0        8.5       431.0  Nintendo      E  
7          58.0          41.0        6.6       129.0  Nintendo      E  
8          87.0          80.0        8.4       594.0  Nintendo      E  
9           NaN           NaN        NaN         NaN       NaN    NaN

## Find the dimensions (number of rows and columns) in the data frame. What do these two numbers represent in the context of the data?

I will use "Shape" on the dataframe to tell us the number of rows and columns. 

In [4]:
Video_Game_Sales.shape

(16719, 16)

As seen above our first number 16,719 is the number of rows of data in the Video_Game_Sales and the second number 16 is the number of columns of data in the dataframe.

## Find the top five games by critic score.

I will use the "nlargest" to find the top five critic scrores

In [13]:
Video_Game_Sales.nlargest(5, ['Critic_Score'])

Name Platform  Year_of_Release     Genre  \
51         Grand Theft Auto IV     X360           2008.0    Action   
57         Grand Theft Auto IV      PS3           2008.0    Action   
227   Tony Hawk's Pro Skater 2       PS           2000.0    Sports   
5350               SoulCalibur       DC           1999.0  Fighting   
16          Grand Theft Auto V      PS3           2013.0    Action   

                 Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  \
51    Take-Two Interactive      6.76      3.07      0.14         1.03   
57    Take-Two Interactive      4.76      3.69      0.44         1.61   
227             Activision      3.05      1.41      0.02         0.20   
5350    Namco Bandai Games      0.00      0.00      0.34         0.00   
16    Take-Two Interactive      7.02      9.09      0.98         3.96   

      Global_Sales  Critic_Score  Critic_Count User_Score  User_Count  \
51           11.01          98.0          86.0        7.9      2951.0   
57           10.50          98.0          64.0        7.5      2833.0   
227           4.68          98.0          19.0        7.7       299.0   
5350          0.34          98.0          24.0        8.8       200.0   
16           21.04          97.0          50.0        8.2      3994.0   

                    Developer Rating  
51             Rockstar North      M  
57             Rockstar North      M  
227   Neversoft Entertainment      T  
5350                    Namco      T  
16             Rockstar North      M

As seen above e have a four way tie for highest critic score of 98. As seen for Grand Theft Auto IV for X360 and PS3, Tony Hawk's Pro Skater 2, and SoulCalibur

## Find the number of video games in the data frame in each genre.

I will use "groupby" to group my genres and sort my values and reset my index for the names of the amount of video games from highest to lowest genre.

In [14]:
Video_Game_Sales.groupby('Genre').size().sort_values(ascending=False).reset_index(name='Name')

Genre  Name
0         Action  3370
1         Sports  2348
2           Misc  1750
3   Role-Playing  1500
4        Shooter  1323
5      Adventure  1303
6         Racing  1249
7       Platform   888
8     Simulation   874
9       Fighting   849
10      Strategy   683
11        Puzzle   580

As seen above the we can see the number of video games per each genre with action being the highest genre with 3370 games.

## Find the first five games in the data frame on the SNES platform.

I will use "loc" on the dataframe for the platform column to determine which rows contain "SNES" and display the first five rows in the dataframe.

In [23]:
Video_Game_Sales.loc[Video_Game_Sales['Platform'].str.contains('SNES')].head(5)

Name Platform  Year_of_Release     Genre  \
18                      Super Mario World     SNES           1990.0  Platform   
56                  Super Mario All-Stars     SNES           1993.0  Platform   
71                    Donkey Kong Country     SNES           1994.0  Platform   
76                       Super Mario Kart     SNES           1992.0    Racing   
137  Street Fighter II: The World Warrior     SNES           1992.0  Fighting   

    Publisher  NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  \
18   Nintendo     12.78      3.75      3.54         0.55         20.61   
56   Nintendo      5.99      2.15      2.12         0.29         10.55   
71   Nintendo      4.36      1.71      3.00         0.23          9.30   
76   Nintendo      3.54      1.24      3.81         0.18          8.76   
137    Capcom      2.47      0.83      2.87         0.12          6.30   

     Critic_Score  Critic_Count User_Score  User_Count Developer Rating  
18            NaN           NaN        NaN         NaN       NaN    NaN  
56            NaN           NaN        NaN         NaN       NaN    NaN  
71            NaN           NaN        NaN         NaN       NaN    NaN  
76            NaN           NaN        NaN         NaN       NaN    NaN  
137           NaN           NaN        NaN         NaN       NaN    NaN

As seen above the first row from the platform column that contained SNES was row 18.

## Find the five publishers with the highest total global sales. Note: You will need to calculate the total global sales for each publisher to do this.

I will use "groupby" to group the publishers and "agg" for my global sales to allow me to sum and reset my index. Once that is completed I can use "nlargest" to display the top five publishers global sales.

In [38]:
sum = Video_Game_Sales.groupby("Publisher").agg({"Global_Sales": "sum"}).reset_index()
sum.nlargest(5, ['Global_Sales'])

Publisher  Global_Sales
361                     Nintendo       1788.81
139              Electronic Arts       1116.96
21                    Activision        731.16
458  Sony Computer Entertainment        606.48
528                      Ubisoft        471.61

As seen above the publisher with the highest total sales was nintendo with 1788.81.

## Create a new column in the data frame that calculates the percentage of global sales from North America. Display the first five rows of the new data frame.

I will create a new column titled "na_percent" for the percentage of North American sales by taking the NA_Sales and dividing that column by the "Global_Sales" column.

In [42]:
Video_Game_Sales['na_percent']=(Video_Game_Sales['NA_Sales']/Video_Game_Sales['Global_Sales'])

In [43]:
Video_Game_Sales.head(5)

Name Platform  Year_of_Release         Genre Publisher  \
0                Wii Sports      Wii           2006.0        Sports  Nintendo   
1         Super Mario Bros.      NES           1985.0      Platform  Nintendo   
2            Mario Kart Wii      Wii           2008.0        Racing  Nintendo   
3         Wii Sports Resort      Wii           2009.0        Sports  Nintendo   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  Critic_Score  \
0     41.36     28.96      3.77         8.45         82.53          76.0   
1     29.08      3.58      6.81         0.77         40.24           NaN   
2     15.68     12.76      3.79         3.29         35.52          82.0   
3     15.61     10.93      3.28         2.95         32.77          80.0   
4     11.27      8.89     10.22         1.00         31.37           NaN   

   Critic_Count User_Score  User_Count Developer Rating  na_percent  
0          51.0          8       322.0  Nintendo      E    0.501151  
1           NaN        NaN         NaN       NaN    NaN    0.722664  
2          73.0        8.3       709.0  Nintendo      E    0.441441  
3          73.0          8       192.0  Nintendo      E    0.476350  
4           NaN        NaN         NaN       NaN    NaN    0.359260

As seen above we have a display of the first five rows with the new "na_percent" column which shows the percentage of sales that come from North America 

## Find the number NaN entries (missing data values) in each column.

I will use "isna" and "sum" to find the number of NaN entries in each column.

In [50]:
Video_Game_Sales['Name'].isna().sum()

2

In [51]:
Video_Game_Sales['Platform'].isna().sum()

0

In [52]:
Video_Game_Sales['Year_of_Release'].isna().sum()

269

In [53]:
Video_Game_Sales['Genre'].isna().sum()

2

In [54]:
Video_Game_Sales['Publisher'].isna().sum()

54

In [55]:
Video_Game_Sales['NA_Sales'].isna().sum()

0

In [56]:
Video_Game_Sales['EU_Sales'].isna().sum()

0

In [57]:
Video_Game_Sales['JP_Sales'].isna().sum()

0

In [58]:
Video_Game_Sales['Other_Sales'].isna().sum()

0

In [59]:
Video_Game_Sales['Global_Sales'].isna().sum()

0

In [60]:
Video_Game_Sales['Critic_Score'].isna().sum()

8582

In [61]:
Video_Game_Sales['Critic_Count'].isna().sum()

8582

In [62]:
Video_Game_Sales['User_Score'].isna().sum()

6704

In [63]:
Video_Game_Sales['User_Count'].isna().sum()

9129

In [64]:
Video_Game_Sales['Developer'].isna().sum()

6623

In [65]:
Video_Game_Sales['Rating'].isna().sum()

6769

In [66]:
Video_Game_Sales['na_percent'].isna().sum()

0

As seen above some columns have a lot of NaN values while some columns have no NaN values.

## Try to calculate the median user score of all the video games. You will likely run into an error because some of the user score entries are a non-numerical string that cannot be converted to a float. Find and replace this string with NaN and then calculate the median. Then, replace all NaN entries in the user score column with the median value.

First I will use the to_numeric approch which will allow me to conver the values in the User_score column into a float. Then using errors='coerce' I will be changing non-numeric values to NaN

In [67]:
Video_Game_Sales['User_Score'] = Video_Game_Sales['User_Score'].apply(pd.to_numeric, errors='coerce')

Next I will replace the NaNs with the median value for the 'User_Score' column. This will be done by using replace on the dataframe and using np.nan and setting the User_Score to the median then setting the regex equal to True.

In [70]:
Video_Game_Sales = Video_Game_Sales.replace(np.nan, Video_Game_Sales['User_Score'].median(), regex=True)

As seen below by doing this we have converted the User_Score column to a float64 that replaced the NaNs with the median from the column.

In [71]:
print(Video_Game_Sales.dtypes)

Name                object
Platform            object
Year_of_Release    float64
Genre               object
Publisher           object
NA_Sales           float64
EU_Sales           float64
JP_Sales           float64
Other_Sales        float64
Global_Sales       float64
Critic_Score       float64
Critic_Count       float64
User_Score         float64
User_Count         float64
Developer           object
Rating              object
na_percent         float64
dtype: object


In [75]:
Video_Game_Sales.User_Score

0        8.0
1        7.5
2        8.3
3        8.0
4        7.5
        ... 
16714    7.5
16715    7.5
16716    7.5
16717    7.5
16718    7.5
Name: User_Score, Length: 16719, dtype: float64

In [74]:
Video_Game_Sales['User_Score'].median()

7.5

As seen above the User_Score column from the Video_Game_Sales data frame has a median of 7.5 and in the output from the User_Score column from the Video_Game_Sales has replaced the NaNs with the median of 7.5